### PR TITLE
Migrate from `Mocha` to `Jest`

### LINKED ISSUE #688: Remove TS version
```ts
// TypeScript Version: 2.8
```
when you have time, please consider updating this, as it has no longer any value here on DT ecosystem, thanks!

_Originally posted by @peterblazejewicz in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62195#discussion_r967804705_

### LINKED ISSUE #705: Maybe migrate from `mySnapshotTest` to Jest + jest-specific-snapshots
Instead of using our own implementation of snapshot testing in `test/restDocs/test.spec.ts` (see `const mySnapshotTest = ...`) use something else.

https://github.com/mochiya98/mocha-chai-jest-snapshot/ is already used in `test/templates.spec.ts`: `.toMatchSnapshot()` but it doesn't allow to separate snapshots into individual files.

Migrating to jest and https://github.com/igor-dv/jest-specific-snapshot would allow that, but wouldn't allow preserving syntax highlight (see https://github.com/igor-dv/jest-specific-snapshot/issues/27)

Need to research our options to see if there are any good tools available that would let us keep our snapshots as is while getting rid of our own implementation.

If there's nothing out there - consider packaging our homegrown snapshot testing utility as a package, or even better yet addressing https://github.com/igor-dv/jest-specific-snapshot/issues/27 and making it available in Jest or Vitest via plugin.

### PR DESCRIPTION
**Pull request by miles170**
Closes #705.

### CODE REVIEW

**Comment by miles170** [comment]
> Sorry for not getting back to you earlier, now that I look at it, it's pretty hard to review changes to snapshots when they are all packed into one file with no syntax highlight... I would like to have them in separate files with proper extensions (to allow syntax highlight). I did a little investigation and didn't find an easy way to do this.
> 
> There's a jest plugin that can help us: https://github.com/igor-dv/jest-specific-snapshot
> 
> However, I find Jest to be significantly slower than Vitest, so if we would migrate from mocha/chai - we probably should migrate to Vitest. And I think that jest-specific-snapshot might work with Vitest since it has a lot of interoperability with Jest, but that needs to be confirmed.
> 
> **TL;DR**, let's investigate how we can keep snapshot files as they currently are, maybe by using Vitest with jest-specific-snapshot plugin
> 
> I appreciate your efforts, and it's unfortunate that I didn't see this through at first! Please feel free to post your findings/design here first before taking the time to code, thank you!

I would like to try separate snapshots in separate files first.
Maybe use the https://github.com/igor-dv/jest-specific-snapshot.

If everything works fine, maybe migrate Jest to Vitest.

**Comment by miles170** [comment]
It seems that even if we separated snapshots into separate files,
we wouldn't get syntax highlighting because snapshots are still stored as strings.

See https://github.com/facebook/jest/issues/2676#issuecomment-749429920

**Comment by Maxim-Mazurok** [comment]
I see... I'd probably suggest creating a separate PR for #688
And then maybe research if there's any better solution for snapshot testing that would preserve syntax highlight in generated files.
Thank you!

**Comment by Maxim-Mazurok** [comment]
I've updated title and description for https://github.com/Maxim-Mazurok/google-api-typings-generator/issues/705 to reflect our findings

**Comment by miles170** [comment]
> I see... I'd probably suggest creating a separate PR for #688 And then maybe research if there's any better solution for snapshot testing that would preserve syntax highlight in generated files. Thank you!

[Jest Snapshot Language Support](https://marketplace.visualstudio.com/items?itemName=tlent.jest-snapshot-language-support) provides syntax highlighting for your Jest snapshots.

**Comment by Maxim-Mazurok** [comment]
Well, it seems to kinda work for d.ts files:
image

But doesn't work well for package.json:
image

or markdown:
image

And also that's not an option when reviewing here on GitHub.

However, with Jest/Vitest snapshots we'll get multiplatform support and all other nice features for free, making this project easier for new contributors, among other things...

I think that if we could at least split templates into individual files, even without syntax highlighting the benefits probably would outweigh the limitations. What do you think?

**Comment by miles170** [comment]
I agree with you. It might be great if we first split snapshot files first, then go on to migrate to vest or syntax highlighting.

**Comment by miles170** [comment]
There is one problem with using `jest-specific-snapshot` that we have to write `ChaiPlugin` for it.

**Comment by Maxim-Mazurok** [comment]
Cool :)
Then let's do this:
- [ ] remove TS Version (https://github.com/Maxim-Mazurok/google-api-typings-generator/issues/688) first in a separate PR, so that I can easily see what changed in old snapshots.
- [ ] Migrate these tests to Jest + jest-specific-snapshots (https://github.com/Maxim-Mazurok/google-api-typings-generator/issues/705)

Thanks!

**Comment by Maxim-Mazurok** [comment]
> There is one problem with using `jest-specific-snapshot` that we have to write `ChaiPlugin` for it.

I don't think it's worth it, we probably can just migrate these tests to jest instead of using mocha-chai-jest-snapshot.

mocha-chai-jest-snapshot is kinda workaround to add snapshot testing for mocha/chai...

**Comment by miles170** [comment]
> > There is one problem with using `jest-specific-snapshot` that we have to write `ChaiPlugin` for it.
> 
> I don't think it's worth it, we probably can just migrate these tests to jest instead of using mocha-chai-jest-snapshot.
> 
> mocha-chai-jest-snapshot is kinda workaround to add snapshot testing for mocha/chai...

I'll see what I can do.

**Comment by miles170** [comment]
Playing around, I found out that jest with typescript did not play well with ES modules.

[TS and ts-jest meet "type": "module"](https://dev.to/antongolub/ts-and-ts-jest-meet-type-module-5chc)

**Comment by Maxim-Mazurok** [comment]
> Playing around, I found out that jest with typescript did not play well with ES modules.
> 
> [TS and ts-jest meet "type": "module"](https://dev.to/antongolub/ts-and-ts-jest-meet-type-module-5chc) [facebook/jest#9430](https://github.com/facebook/jest/issues/9430)

Yeah, I've had a lot of trouble setting up jest with ES Modules, but sometimes I was lucky...
Perhaps we should look into using Vitest straight away since it promises to support ESM out-of-box

**Comment by miles170** [comment]
> > Playing around, I found out that jest with typescript did not play well with ES modules.
> > [TS and ts-jest meet "type": "module"](https://dev.to/antongolub/ts-and-ts-jest-meet-type-module-5chc) [facebook/jest#9430](https://github.com/facebook/jest/issues/9430)
> 
> Yeah, I've had a lot of trouble setting up jest with ES Modules, but sometimes I was lucky... Perhaps we should look into using Vitest straight away since it promises to support ESM out-of-box

I migrated the test to `Jest`. If you wouldn't mind, please check the code. Thanks!

**Comment by miles170** [comment]
The tests passed, but failed at `npm run check`.
Should we exclude `jest.config.js` from linting?

**Comment by miles170** [comment]
By the way, `jest-specific-snapshot` is incompatible with `Jest 29`.

File: test/restDocs/test.spec.ts [hunk 1]
```diff
@@ -1,20 +1,29 @@
-// cspell:word colordiff
-
-import assert from 'node:assert';
-import {execSync, spawnSync} from 'node:child_process';
-import {readFileSync} from 'node:fs';
+import {execSync} from 'node:child_process';
```
Comment (Maxim-Mazurok) [inline test/restDocs/test.spec.ts:1]: Can you please change `execSync` with "rm -rf ${resultFolder}" to use fs.rm() with recursive: true?
https://nodejs.org/api/fs.html#fspromisesrmpath-options
This way we'll hopefully bring back Windows support

File: test/restDocs/test.spec.ts [hunk 2]
```diff
@@ -45,30 +54,11 @@ const mySnapshotTest = async (name: string, action: () => Promise<void>) => {
...
-  }).trim(); // recommended `sudo apt install colordiff`
-  const diffCommand = `${colordiffBinPath} ${snapshotFolder} ${resultFolder}`;
-
-  if (process.argv.includes('--update')) {
-    console.warn(`updating ${name} snapshot...`);
-    execSync(
-      `rm -rf ${snapshotFolder} && mkdir -p ${snapshotFolder} && cp -R ${resultFolder}/. ${snapshotFolder}`
+  readdirSync(resultFolder).forEach(file => {
+    expect(readFileSync(join(resultFolder, file))).toMatchSpecificSnapshot(
+      join(snapshotFolder, file + '.shot')
```
Comment (Maxim-Mazurok) [inline test/restDocs/test.spec.ts:59]: Can we please use `.snap` extension, same as what Jest uses by default?

Comment (Maxim-Mazurok) [reply test/restDocs/test.spec.ts:59]: Also I'd prefer to use template strings instead of concatenation
```ts
`${file}.snap`
```

File: test/restDocs/test.spec.ts [hunk 3]
```diff
@@ -1,20 +1,29 @@
...
 const __dirname = dirname(fileURLToPath(import.meta.url));
 
+expect.extend({
+  toMatchSpecificSnapshot(received, snapshotFile) {
+    return _toMatchSpecificSnapshot.call(
+      this,
+      received.toString(),
+      snapshotFile,
+      this.currentTestName || ''
+    ) as never;
```
Comment (Maxim-Mazurok) [inline test/restDocs/test.spec.ts:19]: ` as never` shouldn't be necessary here.

We have two options:
1. Change `node_modules/@types/jest-specific-snapshot/index.d.ts` using `npx patch-package`
2. Just `import 'jest-specific-snapshot'`

If we go with 2nd option, we'll get:
- slightly worse snapshot names 
- we'll have to add a helper function `const readFileSyncAsUTF8 = (path: string) => readFileSync(path, 'utf-8');`
But on the upside, we will avoid two workarounds:
- no need to patch types pacakge
- no need to override `toMatchSpecificSnapshot` functionality

Both workarounds might cost us in the future as we will likely have to change them as our dependencies update.

So I think it'd be best to use the second approach, what do you think?

If you will go with that approach, you'll need to delete the snapshots folder and re-run tests to generate snapshots with updated names. Thank you!

image

File: test/utils.spec.ts [hunk 4]
```diff
@@ -269,8 +268,8 @@ describe('sleep', () => {
     sleep(1);
     const end = process.hrtime.bigint();
     const sleptForNanoseconds = end - start;
-    assert(900000000 < sleptForNanoseconds); // more than 0.5s
-    assert(2000000000 > sleptForNanoseconds); // less than 2s
+    expect(900000000 < sleptForNanoseconds).toBeTruthy(); // more than 0.5s
```
Comment (Maxim-Mazurok) [inline test/utils.spec.ts:271]: maybe use .toBeLessThan https://jestjs.io/docs/expect#tobelessthannumber--bigint ?

File: test/utils.spec.ts [hunk 5]
```diff
@@ -269,8 +268,8 @@ describe('sleep', () => {
...
+    expect(2000000000 > sleptForNanoseconds).toBeTruthy(); // less than 2s
```
Comment (Maxim-Mazurok) [inline test/utils.spec.ts:272]: maybe use .toBeGreaterThan https://jestjs.io/docs/expect#tobegreaterthannumber--bigint ?

**Review by Maxim-Mazurok — CHANGES_REQUESTED** [review]
Amazing job! I really appreciate your efforts. I've added some comments that I'd like to discuss/address before merging. Thank you!

File: jest.config.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export default {
+  extensionsToTreatAsEsm: ['.ts'],
+  moduleNameMapper: {
+    '^(\\.{1,2}/.*)\\.js$': '$1',
+  },
+  transform: {
+    '^.+\\.ts$': ['ts-jest', {useESM: true}],
```
[the remark this answers is not in the document]
Comment (miles170) [reply jest.config.js:7]: The `ts-jest` presets don't play well with our tests.

File: test/restDocs/test.spec.ts (hunk 2 above)
Comment (miles170) [reply test/restDocs/test.spec.ts:59]: If using `snap` extension, jest will complain that the snapshot is obsolete.

**Comment by miles170** [comment]
I pinned the `ts-jest` version to `29.0.0`. See [[Bug]: Cannot find name 'DepGraphInfo' error after updating to v29.0.1](https://github.com/kulshekhar/ts-jest/issues/3823)

All but two requested changes are resolved.

File: jest.config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (miles170) [reply jest.config.js]: I agree with your point.

File: test/restDocs/test.spec.ts (hunk 2 above)
Comment (miles170) [reply test/restDocs/test.spec.ts]: I have traced to [const testPath = snapshotResolver.resolveTestPath(snapshotFile);](https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/index.ts#L133-L147)
which indicates that the `results` folder should be replaced with `__tests__` and the `snapshots` folder with `__snaphosts__`.

Comment (miles170) [reply test/restDocs/test.spec.ts]: But the compiled JavaScript behavior is different from the Typescript.

Comment (miles170) [reply test/restDocs/test.spec.ts]: We have to use [snapshotResolver](https://jestjs.io/docs/configuration#snapshotresolver-string) to fix this problem.

File: jest.config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Maxim-Mazurok) [reply jest.config.js]: Not that it matters much, but my issue was closed and it looks like I linked to the wrong PR, I think that https://github.com/kulshekhar/ts-jest/pull/3833 will solve both of the issues, would be great to update the readme for future reference (also I appreciate your patience, I don't have time to review the whole thing, I'll try to do it sooner than later, thanks so much!)

File: .eslintignore [hunk 7]
```diff
@@ -2,5 +2,5 @@
 dist/
 types/
 /test/restDocs/*.json
-/test/restDocs/results/
-/test/restDocs/snapshots/
+/test/restDocs/__tests__/
```
Comment (Maxim-Mazurok) [inline .eslintignore:5]: `__tests__` folder [by default](https://jestjs.io/docs/configuration#testmatch-arraystring) is used to find `jest` tests.
But in our case, it looks like it's used to store the generation results, so I got a bit confused by this. The name should better represent the contents of the folder: temporary results of the generator used for tests. This would be a very long name, so probably we can shorten it to "test-results" or "temp-results"?

**Review by Maxim-Mazurok — CHANGES_REQUESTED** [review]
I do appreciate the investigation and effort that went into making sure that snapshots are in individual files, I've never dealt with this before and this is for sure very interesting knowledge about jest's inner workings. Now I can see that this requires much more complication than I originally anticipated. Custom resolver, for example, is very advanced, requires a deep understanding of jest inner-workings, and in case jest breaks backward compatibility in future updates - it'll be very challenging to work with. And we can't use TS yet which isn't good for code quality. And we probably can't really test this complex logic.
Now having this full picture thanks to your research, I think that having one huge file with snapshots isn't that big of a problem. The default approach is quite good for JSON and TS files: it doesn't add extra escaping for double quotes and jest extension in VS Code lets you fold it and does a pretty good job at highlighting. Markdown README will be a challenge as it can't be folded and it takes up many pages so it might be challenging to understand the context, like from which test this part of README is. But that shouldn't be a huge problem.
I'm pretty happy with all the changes in tests. Do you think that it'd make sense to bring all the snapshots back into one file?
Thank you!

File: .eslintignore (hunk 7 above)
Comment (miles170) [reply .eslintignore:5]: Or should it rollback to the original `results`?

File: test/discovery.integration.spec.ts [hunk 8]
```diff
@@ -296,8 +295,14 @@ it('getExtraRestDescriptions works for google ads', async () => {
...
+    'v4',
+    'v5',
+    'v6',
+    'v7',
+    'v8',
+    'v9',
+    'v10',
+    'v11',
+  ]);
+}, 0); // performs requests to the actual server
```
Comment (Maxim-Mazurok) [inline test/discovery.integration.spec.ts:308]: Turns out setting timeout to 0 doesn't disable timeouts in jest, please set it to 30000 instead, should give it enough time. It took just ~2 seconds for CI so that's why it didn't fail, but it tool just a little bit over the default 5s timeout on my computer (probably Australian internet to blame here)

**Review by Maxim-Mazurok — CHANGES_REQUESTED** [review]
I have generated all types using the master branch, committed, then generated using your branch and compared the difference, there was none, except for one package where revision got updated. So I'm quite certain this won't break anything and will be very happy to merge it as soon as you fix that little timeout issue. Also please do not force-push so that I can see what changed. Thank you so much!

File: .eslintignore (hunk 7 above)
Comment (Maxim-Mazurok) [reply .eslintignore]: Yeah, results are good enough, thanks!

**Comment by Maxim-Mazurok** [comment]
Actually, I think it's easier and quicker to just merge it and fix the test myself

**Review by Maxim-Mazurok — APPROVED** [review]

File: test/discovery.integration.spec.ts (hunk 8 above)
Comment (Maxim-Mazurok) [reply test/discovery.integration.spec.ts:308]: Fixed in https://github.com/Maxim-Mazurok/google-api-typings-generator/commit/a613f11cc0607fc2342e895f556f1d1fc1d05d81

**Comment by Maxim-Mazurok** [comment]
Thank you again for an amazing job!
